services/supervisor/src/api.rs

```diff
@@ -100,6 +185,39 @@ impl PrivateApi for ApiImpl<'_> {
         let config = Schema::new(self.0.service_data()).supervisor_config();
         Ok(config)
     }
+
+    fn deploy_status(&self, query: DeployInfoQuery) -> Result<DeployResponse, Self::Error> {
+        let request = DeployRequest::try_from(query)?;
+
+        let schema = Schema::new(self.0.service_data());
+        if let Some(stored_request) = schema.pending_deployments.get(&request.artifact) {
+            // Artifact is currently in the pending deployments, check that request is the same.
+            if request == stored_request {
+                return Ok(DeployState::Pending.into());
+            }
+        }
+        if let Some(height) = schema.deploy_failures.get(&request) {
+            // Request in the list of failed deployments.
+            return Ok(DeployState::Failed(height).into());
+        }
+
+        let dispatcher_schema = self.0.data().for_dispatcher();
+        let core_schema = self.0.data().for_core();
+        let validator_count = core_schema.consensus_config().validator_keys.len();
+        // Check that artifact is deployed and was confirmed *for the same deploy request*.
+        if dispatcher_schema.get_artifact(&request.artifact).is_some()
+            && schema.deploy_confirmations.confirmations(&request) == validator_count
```

Comment:
This looks awkward, TBH. Would it make sense to record the deploy status for each request explicitly in the supervisor schema, rather than just recording failures? If so, this would be a better solution, IMO; it would be more transparent and clean.